### ISSUE TITLE
training & testing time

### ISSUE THREAD

**Issue by KervinBill** [bryanyzhu/Hidden-Two-Stream#2]
Hi, when I training , it takes 2 minutes per iteration, and when testing, the process is slow as well. Can you give me some adivce ? I use one GTX 1080ti.

**Comment by bryanyzhu** [comment]
Hi, for me, i have one Titan X. It usually take me one day to finish the hidden temporal stream training. The process is slow. 

The code itself has data prefetching function. So if the data can fit in your memory, the training should be faster.  For example, the size of UCF101 dataset is around 35GB. If your memory is like 64GB, then after one epoch, the training speed will be much faster. Or put the data on SSD, it also helps.

**Comment by KervinBill** [comment]
Thanks, @bryanyzhu. I put the data on SSD, it is fast now.  You only put the temporal script on github, could you share the fused script ? and I want to know how  you fused the result. I use one script fuse the spatical and temporal net, but got the cuda memory error.

**Comment by bryanyzhu** [comment]
Yes, if you want to fuse them during inference time, you will highly likely to get cuda  memory error. Because both networks are VGG16, and consumes lots of memory. 

What I did is I first get the spatial result ``S``, which is a 101 * 3783 matrix (3783 is the number of test samples in UCF101 split1), and then temporal result ``T``, which is also a 101 * 3783 result. 101 means the dimension of the softmax layer. Save both matrix to hard drive. To fuse them, load them together and simply

``F = S + 2 * T``

You will get the final prediction by doing 

``prediction = argmax(F)``

Since I wrote the fuse script long ago, it is matlab based, so I didn't put it here. If you still have problems, I can email you my matlab scripts and predictions in .mat format.

**Comment by KervinBill** [comment]
Thank you very much. You give me a lot of help. My mail address: zdlyoyo@outlook.com. Look forward to your mail.

**Comment by KervinBill** [comment]
Hi, @bryanyzhu<https://github.com/bryanyzhu>. Thank you very much.   Could you send your matlab scripts and predictions to me ? I fused the predictions, but got a wrong result.



________________________________
发件人: Yi Zhu <notifications@github.com>
发送时间: 2017年12月17日 1:39
收件人: bryanyzhu/Hidden-Two-Stream
抄送: KervinBill; Author
主题: Re: [bryanyzhu/Hidden-Two-Stream] training & testing time (#2)


Yes, if you want to fuse them during inference time, you will highly likely to get cuda memory error. Because both networks are VGG16, and consumes lots of memory.

What I did is I first get the spatial result S, which is a 101 * 3783 matrix (3783 is the number of test samples in UCF101 split1), and then temporal result T, which is also a 101 * 3783 result. 101 means the dimension of the softmax layer. Save both matrix to hard drive. To fuse them, load them together and simply

F = S + 2 * T

You will get the final prediction by doing

prediction = argmax(F)

Since I wrote the fuse script long ago, it is matlab based, so I didn't put it here. If you still have problems, I can email you my matlab scripts and predictions in .mat format.

―

**Comment by bryanyzhu** [comment]
@KervinBill I didn't send it to your email, I updated the repo instead. Please check out the './test' folder for more details. Hope it helps.

**Comment by KervinBill** [comment]
ok, @bryanyzhu , thank you. I will try.

**Comment by SmartPorridge** [comment]
How long it takes you to test the 3783 videos using demo_hidden.py?
It's very very slow in my Tesla P40 GPU...I don't know what's wrong with my settings.

thank you 
@KervinBill 
@bryanyzhu

**Comment by bryanyzhu** [comment]
@SmartPorridge In normal cases, it take about a day. But sometimes it also take 3~4 days on my server. I think it is because the Caffe data loading and system memory stuff, but I am not sure. It didn't happen on my workstation because I use SSD on my workstation but normal HDD on server. There maybe some other reasons like your memory is lower than 64GB, so all the data can't fit in memory and constantly loading lots of images.

**Comment by SmartPorridge** [comment]
@bryanyzhu yeah, I test videos on my server. It also takes 3-4 days..But I have 256G RAM, so I don't think the memory is the bottleneck. Maybe it's because the huge difference between SSD and HDD. 
 I split the videos to 8 parts and test on 8 GPUs , now it takes only half a day. Thank you for your outstanding work!
Best

**Comment by bryanyzhu** [comment]
Glad to hear about it.

**Comment by SmartPorridge** [comment]
@bryanyzhu I have another question, I used your trained model to test UCF-101 split1, the validation accuracy is 73.46% and the eval finally accuracy is 84.78%. It's as same as  your paper. 
  But when I trained my own model finetuning from the model **ucf101_split1_vgg16_init.caffemodel** you provided, I only got a validation accuracy of 59%. I didn't change anything.
  Then I cahnged the batchsize to 16, I got a validation accuracy of 69.87%.
  Could you give me some advice? I don't know why……
Thank you very much!!

**Comment by bryanyzhu** [comment]
Since there are many losses during the network training, you can do a parameter search to get the best results. Another reason maybe the quality of your image is not good. I use cv2 to extract images from video, if you use ffmpeg, then your image quality is lower than mine. Another strategy maybe early stopping, because actually the init model is pre-trained already, it shouldn't take too long to converge. Lowest loss doesn't mean best accuracy. Finally, you can turn off or reduce the loss weights on the flow side to see if you can get a better results.
